webhook/resourcesemantics/conversion/conversion.go

```diff
@@ -0,0 +1,200 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package conversion
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+
+	"go.uber.org/zap"
+	apixv1beta1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"knative.dev/pkg/logging"
+)
+
+func (r *reconciler) Convert(
+	ctx context.Context,
+	req *apixv1beta1.ConversionRequest,
+) *apixv1beta1.ConversionResponse {
+
+	if r.withContext != nil {
+		ctx = r.withContext(ctx)
+	}
+
+	res := &apixv1beta1.ConversionResponse{
+		UID: req.UID,
+		Result: metav1.Status{
+			Status: metav1.StatusSuccess,
+		},
+	}
+
+	result := make([]runtime.RawExtension, 0, len(req.Objects))
+
+	for _, obj := range req.Objects {
+		converted, err := r.convert(ctx, obj, req.DesiredAPIVersion)
+		if err != nil {
+			res.Result.Status = metav1.StatusFailure
+			res.Result.Message = err.Error()
+			break
+		}
+
+		result = append(result, converted)
+	}
+
+	res.ConvertedObjects = result
+	return res
+}
+
+func (r *reconciler) convert(
+	ctx context.Context,
+	inRaw runtime.RawExtension,
+	targetVersion string,
+) (outRaw runtime.RawExtension, err error) {
+
+	logger := logging.FromContext(ctx)
+
+	defer func() {
+		if err != nil {
+			logger.Errorf("Conversion failed: %s", err)
+		}
+	}()
+
+	inGVK, err := parseGVK(inRaw)
+	if err != nil {
+		return
+	}
+
+	logger.Infof("Converting %s to version %s", formatGVK(inGVK), targetVersion)
+
+	inGK := inGVK.GroupKind()
+	conv, ok := r.kinds[inGK]
+	if !ok {
+		err = fmt.Errorf("no conversion support for type %s", formatGK(inGVK.GroupKind()))
+		return
+	}
+
+	outGVK, err := parseAPIVersion(targetVersion, inGK.Kind)
+	if err != nil {
+		return
+	}
+
+	inZygote, ok := conv.Zygotes[inGVK.Version]
+	if !ok {
+		err = fmt.Errorf("conversion not supported for type %s", formatGVK(inGVK))
+		return
+	}
+	outZygote, ok := conv.Zygotes[outGVK.Version]
+	if !ok {
+		err = fmt.Errorf("conversion not supported for type %s", formatGVK(outGVK))
+		return
+	}
+	hubZygote, ok := conv.Zygotes[conv.HubVersion]
+	if !ok {
+		inGK := inGVK.GroupKind()
+		err = fmt.Errorf("conversion not supported for type %s", formatGK(inGK))
+		return
+	}
+
+	in := inZygote.DeepCopyObject().(ConvertibleObject)
+	hub := hubZygote.DeepCopyObject().(ConvertibleObject)
+	out := outZygote.DeepCopyObject().(ConvertibleObject)
+
+	hubGVK := inGVK.GroupKind().WithVersion(conv.HubVersion)
+	logger = logger.With(
+		zap.String("inputType", formatGVK(inGVK)),
+		zap.String("outputType", formatGVK(outGVK)),
+		zap.String("hubType", formatGVK(hubGVK)),
+	)
+
+	// TODO(dprotaso) - potentially error on unknown fields
+	if err = json.Unmarshal(inRaw.Raw, &in); err != nil {
+		err = fmt.Errorf("unable to unmarshal input: %s", err)
+		return
+	}
+
+	if inGVK.Version == conv.HubVersion {
+		hub = in
```

Comment:
@pmorie tells me a useful pattern with conversion webhooks is to convert from one version to the same version in the case of newly added required fields and such. Should we allow that here - converting between two of the same versions? For things like when Serving did the v1alpha1 API changes without changing the version?

Reply:
I'd consider the lemonade migration an exception at the time due to k8s limitations. Also the conversion webhook isn't invoked if the version being served is the same as the storage version.